### PR TITLE
Move openapi common logic to apimachinery

### PR DESCRIPTION
**Pull request by apelisse**
**What this PR does / why we need it**:

Most of the openapi logic has been moved to kube-openapi already, but
this part depends on "GroupVersionKind" that is very apimachinery
specific. Move that code there so it can be used both from the server
and from the client.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by mengqiy** [comment]
Some tests are failing not because of flake.
Ping me for review again when they are green.

File: pkg/proxy/BUILD [hunk 1]
```diff
@@ -26,7 +26,7 @@ filegroup(
     name = "all-srcs",
     srcs = [
         ":package-srcs",
-        "//pkg/proxy/apis:all-srcs",
+        "//pkg/proxy/apis/kubeproxyconfig:all-srcs",
```
Comment (mengqiy) [inline pkg/proxy/BUILD]: Why your PR touch `pkg/proxy`?

**Review by mengqiy — COMMENTED** [review]
Other LG

File: pkg/proxy/BUILD (hunk 1 above)
Comment (apelisse) [reply pkg/proxy/BUILD]: because this was broken, but it was fixed in the mean time in another pr (#55277).

I'll rebase, that should remove that change.

**Comment by apelisse** [comment]
@mengqiy rebased.

**Comment by apelisse** [comment]
cc @caesarxuchao

File: staging/src/k8s.io/apimachinery/Godeps/Godeps.json [hunk 2]
```diff
@@ -115,6 +143,126 @@
 			"Rev": "cca7078d478f8520f85629ad7c68962d31ed7682"
 		},
 		{
+			"ImportPath": "github.com/onsi/ginkgo",
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/Godeps/Godeps.json]: Any chance we can get rid of these depdencies? We want to keep apimachinery's dependency as small as possible.

File: staging/src/k8s.io/apimachinery/Godeps/Godeps.json [hunk 3]
```diff
@@ -217,6 +369,14 @@
 		{
 			"ImportPath": "k8s.io/kube-openapi/pkg/common",
 			"Rev": "39a7bf85c140f972372c2a0d1ee40adbf0c8bfe1"
+		},
+		{
+			"ImportPath": "k8s.io/kube-openapi/pkg/util/proto",
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/Godeps/Godeps.json]: @deads2k @lavalamp any objection to letting k8s.io/apimachinery depends on k8s.io/kube-openapi?

File: staging/src/k8s.io/apimachinery/Godeps/Godeps.json (hunk 2 above)
Comment (apelisse) [reply staging/src/k8s.io/apimachinery/Godeps/Godeps.json]: Yeah, I really don't have a strong feeling about `ginkgo`. Some people really like it, I'm usually also in favor of minimal dependencies.

**Comment by caesarxuchao** [comment]
Per offline chat with @apelisse, currently the only user of the openapi util is `kubectl`, but he thinks in the future apiserver will use these openapi util to do the validation, thus the openapi util should sit in a common library repo rather than in kubectl.

I'm not convinced that apiserver will use openapi to do validation, thus not convinced that we should move the openapi util out of kubectl.

However, I heard CRD is going to use openapi to do validation, is CRD going to use these [util](https://github.com/apelisse/kubernetes/tree/44361f7e12050828230695e976aae20c45a45548/staging/src/k8s.io/apimachinery/pkg/util/openapi) cc @nikhita @sttts? If so, I can accept moving these openapi util to apimachinery.

**Comment by sttts** [comment]
> However, I heard CRD is going to use openapi to do validation, is CRD going to use these util cc @nikhita @sttts? If so, I can accept moving these openapi util to apimachinery.

CRDs use go-openapi for validation already today. 

From a brief look at that util package, I don't see much use of it in apiserver. I would not turn that into a general library before there is any shared use. I.e. I vote against moving it to apimachinery.

**Comment by apelisse** [comment]
I don't really have a strong feeling either way. I've been told that we want more validation on the server, and I could bring that for very low cost.

**Comment by sttts** [comment]
> I don't really have a strong feeling either way. I've been told that we want more validation on the server, and I could bring that for very low cost.

Let's leave it in kubectl until we see a need for sharing it.

**Pull request closed without merging**